File: libs/python/phenotype_sdk/src/phenotype_sdk/adapters/outbound/http_client.py
"""HTTP client adapter for remote configuration service."""

import httpx
from typing import Any, AsyncIterator

from phenotype_sdk.domain.entities import ConfigEntry
from phenotype_sdk.domain.ports import ConfigRepository
from phenotype_sdk.domain.value_objects import ConfigValue, ValueType


class HttpConfigClient(ConfigRepository):
    """
    HTTP adapter for configuration repository.

    Implements the ConfigRepository port (Hexagonal: Driven Adapter).
    Handles HTTP communication with the configuration service.
    """

    def __init__(self, base_url: str, api_key: str | None = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = httpx.AsyncClient(
            headers={"Authorization": f"Bearer {api_key}"} if api_key else {}
        )

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._client.aclose()

    async def get(self, key: str) -> ConfigEntry | None:
        """Retrieve a configuration entry from the remote service."""
        response = await self._client.get(f"{self._base_url}/api/v1/config/{key}")
        if response.status_code == 404:
            return None
        response.raise_for_status()

        data = response.json()
        return self._deserialize_entry(data)

    async def save(self, entry: ConfigEntry) -> ConfigEntry:
        """Save a configuration entry to the remote service."""
        response = await self._client.post(
            f"{self._base_url}/api/v1/config",
            json=self._serialize_entry(entry),
        )
        response.raise_for_status()
        return self._deserialize_entry(response.json())

    async def delete(self, key: str) -> bool:
        """Delete a configuration entry from the remote service."""
        response = await self._client.delete(f"{self._base_url}/api/v1/config/{key}")
        if response.status_code == 404:
            return False
        response.raise_for_status()
        return True

    async def list(self, prefix: str | None = None) -> AsyncIterator[ConfigEntry]:
        """List all configuration entries from the remote service."""
        params = {"prefix": prefix} if prefix else {}
        response = await self._client.get(f"{self._base_url}/api/v1/config", params=params)
        response.raise_for_status()

        for item in response.json()["items"]:
            yield self._deserialize_entry(item)

    @staticmethod
    def _serialize_entry(entry: ConfigEntry) -> dict[str, Any]:
        """Serialize domain entity to JSON for API."""
        return {
            "key": entry.key,
            "value": entry.value.raw,
            "value_type": entry.value.value_type.value,
            "version": entry.version,
            "metadata": entry.metadata,
        }

    @staticmethod
    def _deserialize_entry(data: dict[str, Any]) -> ConfigEntry:
        """Deserialize JSON from API to domain entity."""
        from datetime import datetime
        from uuid import UUID

        return ConfigEntry(
            key=data["key"],
            value=ConfigValue(raw=data["value"], value_type=ValueType(data["value_type"])),
            version=data["version"],
            created_at=datetime.fromisoformat(data["created_at"]),
            updated_at=datetime.fromisoformat(data["updated_at"]),
            metadata=data.get("metadata", {}),
            id=UUID(data["id"]),
        )

File: libs/zig/phenotype-sdk/src/adapters/outbound/http_client.zig
//! HTTP client adapter for remote configuration service
//!
//! Implements ConfigRepository port for HTTP-based remote calls.

const std = @import("std");
const ConfigEntry = @import("domain/entities.zig").ConfigEntry;
const ConfigRepository = @import("domain/ports.zig").ConfigRepository;

/// HTTP-based configuration repository
pub const HttpConfigRepository = struct {
    base_url: []const u8,
    api_key: ?[]const u8,
    client: std.http.Client,
    allocator: std.mem.Allocator,

    /// Create HTTP repository
    pub fn init(base_url: []const u8, api_key: ?[]const u8, allocator: std.mem.Allocator) HttpConfigRepository {
        return .{
            .base_url = base_url,
            .api_key = api_key,
            .client = std.http.Client{ .allocator = allocator },
            .allocator = allocator,
        };
    }

    /// Get configuration entry
    pub fn get(self: *HttpConfigRepository, key: []const u8) !?ConfigEntry {
        const url = try std.fmt.allocPrint(self.allocator, "{s}/api/v1/config/{s}", .{ self.base_url, key });
        defer self.allocator.free(url);

        var request = try self.client.open(.GET, try std.Uri.parse(url), .{});
        defer request.deinit();

        if (self.api_key) |key| {
            try request.headers.setAuthorization(.{ .Bearer = key });
        }

        try request.send(.{});
        try request.wait();

        if (request.response.status == .not_found) {
            return null;
        }

        if (request.response.status != .ok) {
            return error.RequestFailed;
        }

        // Parse response (simplified)
        const body = try request.reader().readAllAlloc(self.allocator, 1024 * 1024);
        defer self.allocator.free(body);

        _ = body;
        return null; // Would parse JSON here
    }

    /// Create port adapter
    pub fn asPort(self: *HttpConfigRepository) ConfigRepository {
        return ConfigRepository{
            .get = getPortWrapper,
            .save = savePortWrapper,
            .delete = deletePortWrapper,
            .list = listPortWrapper,
        };
    }
};

fn getPortWrapper(key: []const u8) !?ConfigEntry {
    _ = key;
    return null;
}

fn savePortWrapper(entry: ConfigEntry) !ConfigEntry {
    _ = entry;
    return entry;
}

fn deletePortWrapper(key: []const u8) !bool {
    _ = key;
    return false;
}

fn listPortWrapper(prefix: ?[]const u8) !std.array_hash_map.ArrayHashMap([]const u8, ConfigEntry) {
    _ = prefix;
    return std.array_hash_map.ArrayHashMap([]const u8, ConfigEntry).init(.{});
}
